subsys/net/ip/tcp2.c

```diff
@@ -832,6 +936,40 @@ static struct tcp *tcp_conn_new(struct net_pkt *pkt)
 	return conn;
 }
 
+static void update_acked_data(struct tcp *conn, struct tcphdr *th)
+{
+	/* If there is any pending data, remove them from re-send as peer
+	 * just acked it.
+	 */
+	u32_t acked_len = th_ack(th) - conn->last_seq;
+	int ret;
+
+	if (!conn->pending_data || acked_len == 0) {
+		return;
+	}
+
+	NET_DBG("Peer acked %u bytes", acked_len);
+
+	net_pkt_cursor_init(conn->pending_data);
+	net_pkt_set_overwrite(conn->pending_data, true);
+
+	ret = net_pkt_pull(conn->pending_data, acked_len);
+	if (ret < 0 && conn->pending_len != (int)acked_len) {
+		NET_ERR("Trying to ack more data (%d) that was in the "
+			"queue (%u)", acked_len, conn->pending_len);
+		conn->pending_len = 0;
+	} else {
+		conn->pending_len -= acked_len;
+		if (conn->pending_len < 0) {
+			NET_DBG("Pending length underflow by %d",
+				conn->pending_len);
+			conn->pending_len = 0;
+		}
+	}
+
+	conn->last_seq += acked_len;
```

Comment:
I am not sure, I follow regarding the purpose of last_seq. Could you please try your changes with the TTCN-3 sanity check?

Reply:
> I am not sure, I follow regarding the purpose of last_seq

I had issue when trying to use conn->seq, it was updated too early to be useful. But after these latest refactorings, perhaps it works now. I will try that.

Reply:
If I remove the conn->last_seq, and use conn->seq directly, then we start to get these kind of issues (if we sent long packet and receiving only ACK to small portition of waiting data):
```
[00:00:29.700,000] <dbg> net_echo_server_sample.handle_data: TCP (IPv6): Received and replied with 640 bytes
[00:00:29.700,000] <dbg> net_tcp.tcp_in: (0x08081090): ACK,PSH Seq=2560866935 Ack=334960246 Len=592 [ESTABLISHED Seq=334960886 Ack=256
[00:00:29.700,000] <dbg> net_tcp.update_acked_data: (0x08081090): Peer acked 4294966657 bytes
[00:00:29.700,000] <err> net_tcp: Trying to ack more data (-639) that was in the queue (671)
```
So the idea of last_seq was to keep track of what is the position in the data stream that we are waiting ack. So in theory we should be able to use conn->seq for that but there is now some complex interaction in the code but I am a bit lost where to look.

Reply:
> So in theory we should be able to use conn->seq for that but there is now some complex interaction in the code but I am a bit lost where to look.

One way would be not to update conn->seq as it’s done in this patch, but to use pending_len to calculate the sequence number for the subsequent data packets.

Reply:
Could be, although everything seems to work now and I would like to avoid this optimization at this point.